frontend/packages/console-shared/src/promql/cluster-dashboard.ts

```diff
@@ -217,15 +217,13 @@ const top25Queries = {
     `
       topk(25, sort_desc(
         count by (namespace) (
-          (
             kube_running_pod_ready
             *
-            on(pod,namespace) group_left(node) (node_namespace_pod:kube_pod_info:)
+            ignoring(node,uid) group_right node_namespace_pod:kube_pod_info:)
```

Comment:
We could also drop this inner `*` as this only serves to join in the node label. We could achieve the same as in the `CPU_UTILIZATION` query by using `label_replace`.

Reply:
The query would look like this:
```
topk(25, sort_desc(
        count by (namespace) (
            kube_running_pod_ready
          )
          *
          on(instance) group_left() 
          (label_replace(max by (node) (kube_node_role{role=~"<%= nodeType %>"}), "instance", "$1", "node","(.*)"))
        )

Reply:
> By the way I'm not sure where this query is used in the console.

@simonpasquier @jan--f This is used in the console when showing the Pod Count Breakdown:

<img width="473" alt="Screenshot 2022-04-21 at 10 16 30" src="https://user-images.githubusercontent.com/5461414/164411564-90db899e-97d2-49f1-9d06-827b64bab004.png">

Reply:
I'll try to explain why the current query may trigger many-to-many errors and why topk() could fix it. Let's focus on this part:

```
kube_running_pod_ready
*
on(pod,namespace) group_left(node) (node_namespace_pod:kube_pod_info:)
```

Assuming that we have the following data for `kube_running_pod_ready` and `node_namespace_pod:kube_pod_info:`:

```
kube_running_pod_ready{namespace="foo", pod="bar-0",uid="abc"}

node_namespace_pod:kube_pod_info:{namespace="foo", node="node-1", pod="bar-0"}
node_namespace_pod:kube_pod_info:{namespace="foo", node="node-2", pod="bar-0"}
```

The duplicate entries for `node_namespace_pod:kube_pod_info:` would likely be a statefulset pod that migrated from one node to another and for a short period of time, kube-state-metrics reports as if the pod was running on the 2 nodes. Executing the PromQL query at this point in time (when 2 series are returned for the same (namespace,pod) tuple) would yield the many-to-many as described in [BZ 2057251](https://bugzilla.redhat.com/show_bug.cgi?id=2057251#c0).

To solve this type of issue, the usual solution is to use an aggregation operator like max() or min(). Because we have a non-empty group_left(), the only operator that would help is `topk by(...) (1, ...)` because it will retain all original labels . From https://prometheus.io/docs/prometheus/latest/querying/operators/#aggregation-operators:

> topk and bottomk are different from other aggregators in that a subset of the input samples, including the original labels, are returned in the result vector. by and without are only used to bucket the input vector.

The trick is to include the labels from the `on()` clause in the topk's `by(...)` clause. Thus `topk by(pod, namespace) (1,node_namespace_pod:kube_pod_info:)` would return only one item (either `node_namespace_pod:kube_pod_info:{namespace="foo", node="node-1", pod="bar-0"}` or `node_namespace_pod:kube_pod_info:{namespace="foo", node="node-2", pod="bar-0"}`) and now we're sure that there will be a 1:1 match between `kube_running_pod_ready` and `node_namespace_pod:kube_pod_info:`.

Reply:
I suspect that `kube_running_pod_ready * ignoring(node,uid) group_right node_namespace_pod:kube_pod_info:)` could be affected by the same "many-to-many" errors if kube_running_pod_ready has 2 entries for the same (pod, namespace) tuple (e.g. with different uid).

Reply:
> I suspect that `kube_running_pod_ready * ignoring(node,uid) group_right node_namespace_pod:kube_pod_info:)` could be affected by the same "many-to-many" errors if kube_running_pod_ready has 2 entries for the same (pod, namespace) tuple (e.g. with different uid).

Right, I think I didn't consider that possibility. Thanks a lot for the detailed explanation. Update coming up.